### PR TITLE
Sales analyst refactor

### PR DESCRIPTION
**Pull request by jrwhitmer**
In this PR, we did the new methods for iteration 2 on the sales analyst. we also created a module for averages, standard deviation, and z-scores to reduce redundancy in our code. Also included in this PR is some refactoring of the way our csvs are called within sales engine in hopes that it would reduce test time. 

Looking into using sample sizes of our data for some of our methods to reduce test time as well.

### CODE REVIEW

File: lib/mathable.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  def average(value, total)
+    value / total
+  end
+
+  def standard_deviation(values, average)
+    squared_differences = values.sum do |value|
+      ((value - average)**2).to_f
+    end
+    divided_sum = ((squared_differences) / (values.length - 1))
```
[the remark this answers is not in the document]
Comment (rcasias) [reply lib/mathable.rb:15]: should we remove the set of parentheses around "squared_difference"?

File: lib/sales_engine.rb [hunk 2]
```diff
@@ -9,9 +9,9 @@ class SalesEngine
                 :invoices
 
   def initialize(csv_files)
-    @items = ItemRepo.new(csv_files[:items], self)
-    @merchants = MerchantRepo.new(csv_files[:merchants], self)
-    @invoices = InvoiceRepo.new(csv_files[:invoices], self)
+    @items = ItemRepo.new(csv_files[:items], self) unless csv_files[:items] == nil
```
Comment (deebot10) [inline lib/sales_engine.rb:12]: I like this method for cutting down out run time and accessing the different repos separate instead of all at one time.

Comment (rcasias) [reply lib/sales_engine.rb:12]: Amazing job and research on the cut time of this method and testing!!!

File: lib/sales_analyst.rb [hunk 3]
```diff
@@ -19,15 +21,31 @@ def initialize(items_repo, merchants_repo, invoices_repo)
   end
 
   def all_items
-    @items_repo.all
+    @all_items ||= @items_repo.all
   end
 
   def all_merchants
-    @merchants_repo.all
+    @all_merchants ||= @merchants_repo.all
```
Comment (deebot10) [inline lib/sales_analyst.rb:28]: These new lines of code are perfect for cutting down our run time

**Review by deebot10 — APPROVED** [review]

**Comment by rcasias** [comment]
Great job on iteration 2! Amazing job cutting times for testing!

File: lib/invoices.rb [hunk 4]
```diff
@@ -13,7 +13,7 @@ def initialize(attributes, repository)
     @id = attributes[:id].to_i
     @customer_id = attributes[:customer_id].to_i
     @merchant_id = attributes[:merchant_id].to_i
-    @status = attributes[:status]
+    @status = attributes[:status].to_sym
```
Comment (ahenzler) [inline lib/invoices.rb:16]: Smart adding this in here.

File: lib/mathable.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module Mathable
+
```
[the remark this answers is not in the document]
Comment (ahenzler) [reply lib/mathable.rb:2]: yeah we need to remove this during the refactor

File: lib/sales_analyst.rb [hunk 6]
```diff
@@ -19,15 +21,31 @@ def initialize(items_repo, merchants_repo, invoices_repo)
   end
 
   def all_items
-    @items_repo.all
```
Comment (ahenzler) [inline lib/sales_analyst.rb:22]: these are really creative and should help with run time significantly

File: lib/sales_analyst.rb [hunk 7]
```diff
@@ -43,39 +61,15 @@ def items_per_merchant
...
-    square_root.round(2)
-  end
-
-  def z_score_items(value)
-    ((value - average_items_per_merchant) / average_items_per_merchant_standard_deviation).to_f
-  end
-
-  def z_score_price(value)
-    ((value - average_average_price_per_merchant) /average_unit_price_standard_deviation).to_f
+    standard_deviation(items_per_merchant, average_items_per_merchant)
```
Comment (ahenzler) [inline lib/sales_analyst.rb:64]: love the use of the module

File: lib/sales_analyst.rb [hunk 8]
```diff
@@ -100,22 +94,18 @@ def merchants_with_high_item_count
...
 
   def average_average_price_per_merchant
     array_of_average_item_prices = all_merchants.map do |merchant|
       merchant_id = merchant.id
       average_item_price_for_merchant(merchant_id)
     end
-    sum = array_of_average_item_prices.sum
-    average = sum / array_of_average_item_prices.length
-    result = BigDecimal(average).round(2)
+   BigDecimal(average(array_of_average_item_prices.sum, array_of_average_item_prices.length)).round(2)
```
[the remark this answers is not in the document]
Comment (ahenzler) [reply lib/sales_analyst.rb:108]: we need to change indentation here when we refactor

File: lib/sales_engine.rb (hunk 2 above)
Comment (ahenzler) [reply lib/sales_engine.rb:12]: the use of unless is exactly what we needed here.

**Review by ahenzler — APPROVED** [review]

**Review by rcasias — APPROVED** [review]

**Pull request merged**